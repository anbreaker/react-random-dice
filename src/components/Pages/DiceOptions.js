import React, { useState } from 'react';
import { isInt } from 'validator';
import { ramdom } from '../../Helpers/numRandom';

import { useForm } from '../../hooks/useForm';

export const DiceOptions = () => {
  const [viewErrorFace, setViewErrorFace] = useState(false);
  const [viewErrorDice, setViewErrorDice] = useState(false);
  const [result, setResult] = useState(1);

  const { formValues, handleInputChange } = useForm({
    numFaces: '',
    numDices: '',
  });

  const { numFaces, numDices } = formValues;

  const validateInput = (event) => {
    event.preventDefault();

    if (isInt(numFaces) && numFaces >= 3) setViewErrorFace(false);
    else setViewErrorFace(true);

    if (isInt(numDices)) setViewErrorDice(false);
    else setViewErrorDice(true);

    rollDices();
  };

  const rollDices = () => {
    let total = 0;

    for (let i = 0; i < numDices; i++) {
      total += ramdom(numFaces);
    }

    console.log({ total }, { result });

    setResult(total);
  };

  const handleViewErrorFace = () => setViewErrorFace(false);

  const handleViewErrorDice = () => setViewErrorDice(false);

  return (
    <>
      <div className="container mt-5">
        <h1>Lanzar Dados ðŸŽ² </h1>

        <div className="row mt-5">
          <div className="col text-center">
            <div className="btn btn-lg badge-info" onClick={validateInput}>
              <i className={`fas fa-dice-d20 icon`}></i>
            </div>
            <div className="col text-center mt-2">
              <div className="text-info">{`Caras dado: ${numFaces}  ` || ''}</div>
            </div>
          </div>

          <div className="col text-center ">
            <div className="btn btn-lg badge-info" onClick={validateInput}>
              <i className={`fas fa-cubes icon`}></i>
            </div>

            <div className="col text-center mt-2">
              <div className="text-info">{`NÃºmero de dados: ${numDices}  ` || ''}</div>
            </div>
          </div>
        </div>
        <div className="col text-center">
          <div className="count-dices">Suma total: {result}</div>
        </div>

        <div className="jumbotron mt-5">
          <form onSubmit={validateInput}>
            <div className="row">
              <div className="col">
                <div className="form-group">
                  <label htmlFor="numFaceDice">NÃºmero de Caras del Dado:</label>
                  <input
                    type="number"
                    className="form-control"
                    placeholder="Introduce un NÃºmero de Caras para el Dado."
                    name="numFaces"
                    value={numFaces}
                    onChange={handleInputChange}
                    onFocus={handleViewErrorFace}
                  />
                </div>

                {viewErrorFace && (
                  <p className="text-warning">
                    El nÃºmero de caras para el dado introducido, ha de ser un valor entero
                    (n) y mayor de 3. (Si no puedes jugar con monedas ðŸ¤ª)
                  </p>
                )}
              </div>
              <div className="col">
                <div className="form-group">
                  <label htmlFor="nunDices">NÃºmero de Dados:</label>
                  <input
                    type="number"
                    className="form-control"
                    placeholder="Introduce el NÃºmero de Dados que deseas lanzar."
                    name="numDices"
                    value={numDices}
                    onChange={handleInputChange}
                    onFocus={handleViewErrorDice}
                  />
                </div>

                {viewErrorDice && (
                  <p className="text-warning">
                    El nÃºmero de dados introducido, ha de ser un valor entero.
                  </p>
                )}
              </div>
            </div>

            <div className="row justify-content-center">
              <button type="submit" className="btn btn-lg btn-primary">
                Lanzar
              </button>
            </div>
          </form>
        </div>
      </div>
    </>
  );
};
