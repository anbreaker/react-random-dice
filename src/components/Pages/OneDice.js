import React, { useState } from 'react';

import { Jumbo } from '../Parts/Jumbo';

export const OneDice = () => {
  const caras = ['one', 'two', 'three', 'four', 'five', 'six'];

  const [randomNumDado1, setRandomNumDado1] = useState(1);
  const [shake, setShake] = useState(false);

  const ramdom = (max = 6, min = 1) => {
    return Math.floor(Math.random() * (max - min + 1) + 1);
  };

  const handleRollDice1 = () => {
    const valueDiceRandom = ramdom();

    if (randomNumDado1 !== valueDiceRandom) {
      setRandomNumDado1(valueDiceRandom);
      setShake(true);
    } else {
      handleRollDice1();
      setShake(false);
    }
  };

  console.log(shake);

  return (
    <>
      <div className="container mt-5 ">
        <div className="row mt-5">
          <div className="col text-center">
            <div className="btn btn-lg btn-danger" onClick={handleRollDice1}>
              <i
                className={`fas fa-dice-${caras[randomNumDado1 - 1]} icon`}
                // El css anclado directo es simplemente para mostrar que conozco mas formas.
                // style={{ fontSize: '9rem' }}
              ></i>
            </div>
          </div>

          <div className="col text-center">
            <div className="title">{randomNumDado1}</div>
          </div>
        </div>
        <Jumbo
          title={<h4>Uso</h4>}
          text={
            <li>
              Pulsa sobre el dado y saldrá un número aleatorio, pero nunca saldrá dos
              veces seguidas el mismo valor.
            </li>
          }
        />
      </div>
    </>
  );
};
