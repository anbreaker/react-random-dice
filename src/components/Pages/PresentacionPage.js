import React from 'react';

export const PresentacionPage = () => {
  return (
    <>
      <div className="container mt-3">
        <h1>BeKind Software, prueba Técnica</h1>

        <div className="jumbotron mt-2">
          <h4>
            <b>Reto</b>
          </h4>
          <ul>
            <li>Crear una aplicación que permita al usuario "lanzar dados” 🎲. ✔️</li>
            <li>
              Enlace al proyecto en github:
              <a
                href="https://github.com/anbreaker/react-random-dice"
                target="_blank"
                rel="noopener noreferrer">
                {' '}
                BeKind Software, prueba Técnica ✔️
              </a>
            </li>
          </ul>

          <h5>Historias de usuario:</h5>
          <ul>
            <li>Los usuarios deberán poder realizar la acción de lanzar dados. ✔️</li>
            <li>
              Un dado no debe caer dos veces consecutivas en el mismo valor pero mantener
              la aleatoreidad en el resto de las caras. ✔️
            </li>
          </ul>

          <h5>Extras:</h5>
          <p>
            Cumplir con alguna, varias o todas estas funcionalidades extra te da puntos 🙂
          </p>
          <ul>
            <li>Poder elegir la cantidad de caras del dado. ✔️</li>
            <li>Poder tirar múltiples dados a la vez. ✔️</li>
            <li>Poder tirar múltiples dados de diferentes caras a la vez.</li>
            <li>
              Poder ingresar múltiples usuarios y llevar turnos para tirar los dados así
              como un conteo de los mismos.
            </li>
          </ul>

          <h5>Tecnología:</h5>
          <ul>
            <li>El demo debe correr en el navegador y tener una interfaz gráfica ✔️</li>
            <li>
              Puedes usar cualquier librería de javascript (react, vue, svelte, etc) ✔️
            </li>
            <li>Poder tirar múltiples dados de diferentes caras a la vez.</li>
            <li>El proyecto debe estar versionado en git.</li>
          </ul>

          <h5>Presentación:</h5>
          <ul>
            <li>
              El repositorio así como instrucciones para correrlo deben estar en un
              repositorio de github. ✔️
            </li>
            <li>La aplicación debe poder correr en un navegador. ✔️</li>
            <ul>
              <dd>
                Puntos extra si nos mandas el código corriendo en algún servidor. ✔️
              </dd>
            </ul>
          </ul>
        </div>
      </div>
    </>
  );
};
