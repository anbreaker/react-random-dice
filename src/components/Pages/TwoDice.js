import React, { useState } from 'react';

export const TwoDice = () => {
  const caras = ['one', 'two', 'three', 'four', 'five', 'six'];

  const [randomNumDado1, setRandomNumDado1] = useState(1);
  const [randomNumDado2, setRandomNumDado2] = useState(1);

  const ramdom = (max = 6, min = 1) => {
    return Math.floor(Math.random() * (max - min + 1) + 1);
  };

  const handleRollDice1 = () => {
    const valueDiceRandom = ramdom();

    if (randomNumDado1 !== valueDiceRandom) setRandomNumDado1(valueDiceRandom);
    else handleRollDice1();
  };

  const handleRollDice2 = () => {
    const valueDiceRandom2 = ramdom();

    if (randomNumDado2 !== valueDiceRandom2) setRandomNumDado2(valueDiceRandom2);
    else handleRollDice2();
  };

  return (
    <>
      <div className="container mt-5 ">
        <div className="row mt-5">
          <div className="col text-center">
            <div className="btn btn-lg btn-danger" onClick={handleRollDice1}>
              <i
                className={`fas fa-dice-${caras[randomNumDado1 - 1]} icon`}
                // El css anclado directo es simplemente para mostrar que conozco mas formas.
                // style={{ fontSize: '9rem' }}
              ></i>
            </div>
          </div>

          <div className="col text-center">
            <div className="btn btn-lg btn-danger" onClick={handleRollDice2}>
              <i className={`fas fa-dice-${caras[randomNumDado2 - 1]} icon`}></i>
            </div>
          </div>
        </div>
        <div className="col text-center">
          <div className="title">{randomNumDado1 + randomNumDado2}</div>
        </div>
      </div>
    </>
  );
};
