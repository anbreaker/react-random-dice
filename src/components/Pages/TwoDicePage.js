import React, { useState } from 'react';

import { ramdom } from '../../Helpers/numRandom';

import { Jumbo } from '../Parts/Jumbo';

export const TwoDicePage = () => {
  const caras = ['one', 'two', 'three', 'four', 'five', 'six'];

  const [randomNumDado1, setRandomNumDado1] = useState(1);
  const [randomNumDado2, setRandomNumDado2] = useState(1);

  const handleRollDice1 = () => {
    const valueDiceRandom = ramdom();

    if (randomNumDado1 !== valueDiceRandom) setRandomNumDado1(valueDiceRandom);
    else handleRollDice1();
  };

  const handleRollDice2 = () => {
    const valueDiceRandom2 = ramdom();

    if (randomNumDado2 !== valueDiceRandom2) setRandomNumDado2(valueDiceRandom2);
    else handleRollDice2();
  };

  const handleRollDices = () => {
    handleRollDice1();
    handleRollDice2();
  };

  return (
    <>
      <div className="container mt-5 ">
        <div className="row mt-5">
          <div className="col text-center">
            <div className="btn btn-lg btn-danger" onClick={handleRollDice1}>
              <i
                className={`fas fa-dice-${caras[randomNumDado1 - 1]} icon`}
                // El css anclado directo es simplemente para mostrar que conozco mas formas.
                // style={{ fontSize: '9rem' }}
              ></i>
            </div>
          </div>

          <div className="col d-flex align-items-center justify-content-center">
            <div className="btn btn-lg btn-success" onClick={handleRollDices}>
              <i className="fas fa-random"></i>
            </div>
          </div>

          <div className="col text-center">
            <div className="btn btn-lg btn-danger" onClick={handleRollDice2}>
              <i className={`fas fa-dice-${caras[randomNumDado2 - 1]} icon`}></i>
            </div>
          </div>
        </div>
        <div className="col text-center">
          <div className="title">{randomNumDado1 + randomNumDado2}</div>
        </div>

        <Jumbo
          title={<h4>Uso:</h4>}
          text={
            <>
              <li>
                Los dados funcionan individualmente clicando sobre cada uno de ellos, o en
                conjunto, haciendo click sobre el boton central.
              </li>
              <li>Nunca saldr√° dos veces seguidas el mismo valor en el mismo dado</li>
            </>
          }
        />
      </div>
    </>
  );
};
