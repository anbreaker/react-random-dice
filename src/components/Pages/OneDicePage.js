import React from 'react';

import { OneDice } from '../Parts/OneDice';
import { Jumbo } from '../Parts/Jumbo';

export const OneDicePage = () => {
  return (
    <>
      <div className="container mt-5 ">
        <OneDice />

        <Jumbo
          title={<h4>Uso</h4>}
          text={
            <li>
              Pulsa sobre el dado y saldrá un número aleatorio, pero nunca saldrá dos
              veces seguidas el mismo valor.
            </li>
          }
        />
      </div>
    </>
  );
};
