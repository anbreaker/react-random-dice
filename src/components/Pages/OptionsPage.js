import React, { useEffect, useState } from 'react';

import { OneDice } from '../Parts/OneDice';

// import { Jumbo } from './Parts/Jumbo';

export const OptionsPage = () => {
  const [numeroCaras, setNumeroCaras] = useState(3);
  const [numeroDados, setNumeroDados] = useState(1);

  const [rigthNumberFace, setRigthNumberFace] = useState(false);
  const [rigthNumberDices, setRigthNumberDices] = useState(false);

  const handleInputFaces = ({ target }) => {
    const numCaras = parseInt(target.value);

    if (Number.isInteger(numCaras)) {
      setNumeroCaras(target.value);
      setRigthNumberFace(true);
    } else {
      setRigthNumberFace(false);
    }
  };

  const handleInputDices = ({ target }) => {
    setNumeroDados(target.value);
  };

  useEffect(() => {
    if (!validateInputFaces(numeroCaras)) setRigthNumberFace(true);

    if (!validateInputDices(numeroDados)) setRigthNumberDices(true);
  }, [numeroCaras, numeroDados]);

  const validateInputFaces = (numeroCaras) => {
    const numCaras = parseInt(numeroCaras);

    console.log(Number.isInteger(numCaras), numCaras, '<-- numCaras');
  };

  const validateInputDices = (numeroDados) => {
    const numDados = parseInt(numeroDados);

    console.log(Number.isInteger(numDados), numeroDados, '<-- numeroDados');
  };

  return (
    <>
      <div className="container mt-5">
        <h1>Lanzar Dados ðŸŽ² </h1>

        <OneDice />

        <div className="jumbotron">
          <form>
            <div className="row mt-lg-3">
              <div className="col-6">
                <div className="form-group">
                  <label htmlFor="numFaceDice">NÃºmero de Caras del Dado:</label>
                  <input
                    type="number"
                    pattern="\d+"
                    className="form-control"
                    id="numeroCaras"
                    aria-describedby="numeroCarasDado"
                    placeholder="Introduce un NÃºmero de Caras para el Dado."
                    value={numeroCaras}
                    onChange={handleInputFaces}
                    // onFocus={}
                  />
                </div>
              </div>

              <div className="col-6">
                <div className="form-group">
                  <label htmlFor="nunDices">NÃºmero de Dados:</label>
                  <input
                    type="number"
                    pattern="\d+"
                    className="form-control"
                    id="numeroDados"
                    aria-describedby="numeroDados"
                    placeholder="Introduce el NÃºmero de Dados que deseas lanzar."
                    value={numeroDados}
                    onChange={handleInputDices}
                    // onFocus={}
                  />
                </div>
              </div>
              {rigthNumberFace && (
                <div className="col mt-lg-3 justify-content-center">
                  <p className="text-warning text-center">
                    El nÃºmero del dado introducidos ha de ser un numero entero (n) y mayor
                    de 3. (Si no puedes jugar con monedas ðŸ¤ª)
                  </p>
                </div>
              )}

              {rigthNumberDices && (
                <div className="col mt-lg-3 justify-content-center">
                  <p className="text-warning text-center">
                    El nÃºmero de dados introducidos ha de ser un numero entero (n) y mayor
                    de 1.
                  </p>
                </div>
              )}
            </div>

            <div className="row mt-lg-3 justify-content-center">
              <button type="submit" className="btn btn-lg btn-primary">
                Lanzar
              </button>
            </div>
          </form>
        </div>
      </div>
    </>
  );
};
