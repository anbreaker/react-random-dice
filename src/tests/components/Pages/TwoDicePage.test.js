import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import { TwoDicePage } from '../../../components/Pages/TwoDicePage';

describe('Prueba en <TwoDicePage /> ', () => {
  test('Debe mostrar la <TwoDicePage /> correctamente', () => {
    const wrapper = shallow(<TwoDicePage />);

    expect(wrapper).toMatchSnapshot();
  });
});
