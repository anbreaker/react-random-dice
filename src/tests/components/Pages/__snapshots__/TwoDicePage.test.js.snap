// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <TwoDicePage />  Debe mostrar la <TwoDicePage /> correctamente 1`] = `
<Fragment>
  <div
    className="container mt-5 "
  >
    <div
      className="row mt-5"
    >
      <div
        className="col text-center"
      >
        <div
          className="btn btn-lg btn-danger"
          onClick={[Function]}
        >
          <i
            className="fas fa-dice-one icon"
          />
        </div>
      </div>
      <div
        className="col d-flex align-items-center justify-content-center"
      >
        <div
          className="btn btn-lg btn-success"
          onClick={[Function]}
        >
          <i
            className="fas fa-random"
          />
        </div>
      </div>
      <div
        className="col text-center"
      >
        <div
          className="btn btn-lg btn-danger"
          onClick={[Function]}
        >
          <i
            className="fas fa-dice-one icon"
          />
        </div>
      </div>
    </div>
    <div
      className="col text-center"
    >
      <div
        className="title"
      >
        2
      </div>
    </div>
    <Jumbo
      text={
        <React.Fragment>
          <li>
            Los dados funcionan individualmente clicando sobre cada uno de ellos, o en conjunto, haciendo click sobre el boton central.
          </li>
          <li>
            Nunca saldr√° dos veces seguidas el mismo valor en el mismo dado
          </li>
        </React.Fragment>
      }
      title={
        <h4>
          Uso:
        </h4>
      }
    />
  </div>
</Fragment>
`;
